Library/Homebrew/extend/os/mac/requirements/x11_requirement.rb

```diff
@@ -0,0 +1,49 @@
+require "requirement"
+
+Object.send(:remove_const, :X11Requirement)
+
+class X11Requirement < Requirement
+  include Comparable
+  attr_reader :min_version
+
+  fatal true
+  cask "xquartz"
+  download "https://xquartz.macosforge.org"
+
+  env { ENV.x11 }
+
+  def initialize(name = "x11", tags = [])
+    @name = name
+    if /(\d\.)+\d/ =~ tags.first
+      @min_version = Version.create(tags.shift)
```

Comment:
Do XQuartz versions map to X11/Xorg versions at all? Would be good to share the version logic if possible.

Reply:
The `About` window in my xquartz says: "XQuartz 2.7.11 (xorg-server 1.18.4)"

Reply:
We haven't yet found a need in Linuxbrew for a version check in the X11Requirement. This code is here primarily so that this Requirement exposes the same version API as the XQuartz Requirement, so that the tests pass. I can instead remove this version logic and make that particular test (just one) `:needs_macos`.

Reply:
The point is when it's satisfied by something other than the default formula (which will be forced to be up to date regardless), which I'm assuming remains untested.

Reply:
Before the `xorg` formula existed, I used Linuxbrew with the host's X11 implementation. When installing a bottle that depends on `:x11`, the `default_formula` is used. The host's Xorg may be used when installing something from source that depends on `:x11`.

Reply:
That's the case where there should be a version check.

Reply:
I haven't run into a failure in the wild where the host X11 was too old. I'd rather not fix problems that haven't been encountered by users.

Reply:
The lack of proper version handling is why the Java requirement is currently broken on Linux. I don't think we should repeat that again.

Reply:
```
❯❯❯ grep -r 'depends_on :x11 =>.*[0-9]' .
./gv.rb:  depends_on :x11 => "2.7.2"
```
Is `gv` the only formula that has a minimum version requirement for X11?

Reply:
The fact that they all don't have it is basically a bug. The bottles usually break when they were built with a different version from what the user has. We even discussed disabling all bottles on formula with an x11 requirement at one point because of that.

Reply:
We haven't encountered that issue with Xorg on Linux. When the user installs a bottle that depends on `:x11`, it uses the `xorg` formula, which upgrades that formula.

Reply:
Same behavior for ruby and perl and yet we have version checks.

Reply:
What is the downside to having a feature complete implementation of the requirement?

Reply:
It's unnecessary code if it solves a hypothetical problem.

Reply:
And yet here we have the broken Java requirement, now with a stale label, https://github.com/Linuxbrew/brew/issues/509.

Reply:
Java has 168 formulae with a versioned dependency. X11 has one. These are not comparable.
```
❯❯❯ grep -r 'depends_on :java => "[0-9]' $(brew --repo homebrew/core) | wc -l
     168
```

Reply:
That's because we've set a lax >= on those as a matter of course.

Reply:
More generally, I don't like the continuing precedent this is setting. macOS has a requirement with a version check, and Linuxbrew has a default formula without a version check until such time as something breaks, incurring a technical debt and burden shifting the completion of the work to the future.

Reply:
Also, given over 40% of the x11 requirements are optional, meaning the dependent's bottle won't be used, it seems particularly relevant to have some sort of version check on the system x11 before brew decides the default formula shouldn't be used to satisfy the requirement after all.

Reply:
Note, too, that `check_xquartz_up_to_date` (https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/diagnostic.rb#L290-L299) warns about outdated XQuartz on macOS, which likely papers over some of the issues with older versions.

Reply:
I agree it would be good to have a version check here.

Reply:
I've added a version check that `xdpyinfo -version` is at least version 1.3.1. The most recent version is 1.3.2. I tested that `gv` works on CentOS 6 with the host's Xorg (with `xdpyinfo -version` of 1.3.1) using `brew install -s gv`.

Reply:
How does that correspond to e.g. the X.org 7.7 release: https://www.x.org/wiki/Releases/ ?

Reply:
Looks like checking `/usr/bin/X --version` may be the most authoritative with that in mind.

Reply:
Headless Linux servers don't usually have the X server installed (`/usr/bin/X`), just the X client. The X server is run on your local machine (which is not necessarily running Linuxbrew), whereas the X client is run on the headless server (which is running Linuxbrew). For this reason, I'm checking the version of a X client executable (`/usr/bin/xdpyinfo`) rather than the X server.

The most recent complete release of Xorg X11R7.7 shipped with `xdpyinfo 1.3.0`.
The previous release of Xorg X11R7.6 shipped with `xdpyinfo 1.2.0`.
See https://www.x.org/releases/X11R7.7/changelog.html

Reply:
That's the `xdpyinfo` version rather than the X11 version, though; it's not particularly useful in this case. Can you see if there's anything else that provides the actual X11 version even with the X server not installed? If not, I think this should just assume the server is not headless and the Linux requirement can always override that.

Reply:
The katamari 7.7 had server version 1.12.2 and xdpyinfo 1.3.0. So I think starting with `Xorg -version` >= 1.12.2 makes sense, and then if the `Xorg` command isn't available, fall back on `xdpyinfo -version` > = 1.3.0.

Linux:
```
openzfs ~ # Xorg -version 2>&1|grep Server
X.Org X Server 1.19.3
openzfs ~ # xdpyinfo -version
xdpyinfo 1.3.2
```

macOS with XQuartz `2.7.11`:
```
iMac-TMP:~ joe$ Xorg -version 2>&1|grep Server
X.Org X Server 1.18.4
iMac-TMP:~ joe$ xdpyinfo -version
xdpyinfo 1.3.2
```

Reply:
@sjackman are you planning on actually using the `satisfy` block from this upstreamed code or just continuing to use
```
  satisfy build_env: false do
    Formula["linuxbrew/xorg/xorg"].installed?
  end
```
in Linuxbrew itself?

Reply:
Yes, once it's merged into Homebrew/brew, and then after that, merged into Linuxbrew/brew.